### PR TITLE
[Merged by Bors] - feat(Topology/Algebra/Order): continuous injective function on interval is strictly monotone

### PR DESCRIPTION
**Pull request by shuxuezhuyi**
Suppose `f : [a, b] → δ` or `f : (a, b) → δ` is continuous and injective. Then `f` is strictly monotone.

Co-authored-by: Johan Commelin <johan@commelin.net>

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Algebra/Order/IntermediateValue.lean [hunk 1]
```diff
@@ -644,3 +644,152 @@ theorem ContinuousOn.surjOn_of_tendsto' {f : α → δ} {s : Set α} [OrdConnect
...
+      have p3 : ContinuousOn f (Icc s t) := ContinuousOn.mono p1 h12
+      push_neg at h7
+      have h13 := intermediate_value_Ioo' (le_of_lt h3) p3
+      have h14 : f a ∈ Ioo (f t) (f s) := ⟨h11, h7⟩
+      choose u hu using h13 h14
+      have h15 : u = a :=
+        p2 (h12 (Ioo_subset_Icc_self hu.1)) (left_mem_Icc.mpr (le_of_lt h1)) hu.2
+      have h16 : a < u := lt_of_lt_of_le' hu.1.1 hs.1
+      rw [h15] at h16
+      exact LT.lt.false h16
```
Comment (jcommelin) [inline Mathlib/Topology/Algebra/Order/IntermediateValue.lean:704]:
```suggestion
/-- If `f : [a, b] → δ` is continuous and injective, then `f` is strictly monotone. -/
theorem ContinuousOn.StrictMonoOn_of_InjOn {a b : α} {f : α → δ} (hab : a ≤ b) (hfab : f a ≤ f b)
    (hf_c : ContinuousOn f (Icc a b)) (hf_i : InjOn f (Icc a b)) :
    StrictMonoOn f (Icc a b) := by
-- I learned this proof from
-- <https://www.math.cuhk.edu.hk/course_builder/2021/math1050b/1050b-l14h-3.pdf>
  intro s hs t ht hst
  by_contra' hfts
  replace hfts : f t < f s := lt_of_le_of_ne hfts <| hf_i.ne ht hs hst.ne'
  by_cases h7 : f s ≤ f a
  · have h9 : Icc t b ⊆ Icc a b := Icc_subset_Icc_left ht.1
    have h11 : f a ∈ Ioc (f t) (f b) := ⟨(hfts.trans_le h7), hfab⟩
    obtain ⟨u, hu⟩ : f a ∈ f '' Ioc t b :=
      intermediate_value_Ioc ht.2 (hf_c.mono h9) h11
    apply (lt_of_lt_of_le' hu.1.1 ht.1).ne'
    exact hf_i (h9 (Ioc_subset_Icc_self hu.1)) (left_mem_Icc.mpr hab) hu.2
  · by_cases h10 : f a < f t
    · have h11 : Icc a s ⊆ Icc a b := Icc_subset_Icc_right hs.2
      obtain ⟨u, hu⟩ : f t ∈ f '' Ioo a s :=
        intermediate_value_Ioo hs.1 (hf_c.mono h11) ⟨h10, hfts⟩
      apply (hu.1.2.trans hst).ne
      exact hf_i (h11 (Ioo_subset_Icc_self hu.1)) ht hu.2
    · push_neg at h7 h10
      have h8 : a ≠ t := (lt_of_lt_of_le' hst hs.1).ne
      have h9 : f a ≠ f t := hf_i.ne (left_mem_Icc.mpr hab) ht h8
      have h11 : f t < f a := h9.symm.lt_of_le h10
      have h12 : Icc s t ⊆ Icc a b := Icc_subset_Icc hs.1 ht.2
      obtain ⟨u, hu⟩ : f a ∈ f '' Ioo s t :=
        intermediate_value_Ioo' hst.le (hf_c.mono h12) ⟨h11, h7⟩
      apply (lt_of_lt_of_le' hu.1.1 hs.1).ne'
      exact hf_i (h12 (Ioo_subset_Icc_self hu.1)) (left_mem_Icc.mpr hab) hu.2
```

**Review by jcommelin — COMMENTED** [review]
Thanks for your PR! It seems that you are sometimes missing a few tricks.

I had a go at optimizing and generalizing your PR a bit.
You should try and see if you can apply more of my tricks to clean it up further.

Here is the resulting code:
```
/-- If `f : [a, b] → δ` is continuous and injective, then `f` is strictly monotone. -/
theorem ContinuousOn.StrictMonoOn_of_InjOn {a b : α} {f : α → δ} (hab : a ≤ b) (hfab : f a ≤ f b)
    (hf_c : ContinuousOn f (Icc a b)) (hf_i : InjOn f (Icc a b)) :
    StrictMonoOn f (Icc a b) := by
-- I learned this proof from
-- <https://www.math.cuhk.edu.hk/course_builder/2021/math1050b/1050b-l14h-3.pdf>
  intro s hs t ht hst
  by_contra' hfts
  replace hfts : f t < f s := lt_of_le_of_ne hfts <| hf_i.ne ht hs hst.ne'
  by_cases h7 : f s ≤ f a
  · have h9 : Icc t b ⊆ Icc a b := Icc_subset_Icc_left ht.1
    have h11 : f a ∈ Ioc (f t) (f b) := ⟨(hfts.trans_le h7), hfab⟩
    obtain ⟨u, hu⟩ : f a ∈ f '' Ioc t b :=
      intermediate_value_Ioc ht.2 (hf_c.mono h9) h11
    apply (lt_of_lt_of_le' hu.1.1 ht.1).ne'
    exact hf_i (h9 (Ioc_subset_Icc_self hu.1)) (left_mem_Icc.mpr hab) hu.2
  · by_cases h10 : f a < f t
    · have h11 : Icc a s ⊆ Icc a b := Icc_subset_Icc_right hs.2
      obtain ⟨u, hu⟩ : f t ∈ f '' Ioo a s :=
        intermediate_value_Ioo hs.1 (hf_c.mono h11) ⟨h10, hfts⟩
      apply (hu.1.2.trans hst).ne
      exact hf_i (h11 (Ioo_subset_Icc_self hu.1)) ht hu.2
    · push_neg at h7 h10
      have h8 : a ≠ t := (lt_of_lt_of_le' hst hs.1).ne
      have h9 : f a ≠ f t := hf_i.ne (left_mem_Icc.mpr hab) ht h8
      have h11 : f t < f a := h9.symm.lt_of_le h10
      have h12 : Icc s t ⊆ Icc a b := Icc_subset_Icc hs.1 ht.2
      obtain ⟨u, hu⟩ : f a ∈ f '' Ioo s t :=
        intermediate_value_Ioo' hst.le (hf_c.mono h12) ⟨h11, h7⟩
      apply (lt_of_lt_of_le' hu.1.1 hs.1).ne'
      exact hf_i (h12 (Ioo_subset_Icc_self hu.1)) (left_mem_Icc.mpr hab) hu.2

/-- If `f : [a, b] → δ` is continuous and injective, then `f` is strictly antitone. -/
theorem ContinuousOn.StrictAntiOn_of_InjOn {a b : α} {f : α → δ} (h1 : a ≤ b) (hfab : f b ≤ f a)
    (p1 : ContinuousOn f (Icc a b)) (hf_i : InjOn f (Icc a b)) :
    StrictAntiOn f (Icc a b) := by
    let g (x : α) : OrderDual δ := f x
    have h2' : g a ≤ g b := hfab
    exact ContinuousOn.StrictMonoOn_of_InjOn h1 h2' p1 hf_i

/-- Suppose f : [a, b] → δ is continuous and injective. Then f is strictly monotone.-/
theorem ContinuousOn.StrictMonoOn_of_InjOn'' {a b : α} {f : α → δ} (h1 : a ≤ b)
    (p1 : ContinuousOn f (Icc a b)) (p2 : InjOn f (Icc a b)) :
    StrictMonoOn f (Icc a b) ∨ StrictAntiOn f (Icc a b) :=
  (le_total (f a) (f b)).imp
    (ContinuousOn.StrictMonoOn_of_InjOn h1 · p1 p2)
    (ContinuousOn.StrictAntiOn_of_InjOn h1 · p1 p2)

/-- This lemma is only used to
prove and will be immediately superseded by
`continuous_inj_is_monotone_Ioo` below.
It asserts that a continuous injective `f: (a, b) → δ` must be strictly monotone on
every closed interval `[c, d] ⊆ (a, b)`.-/
private lemma continuous_inj_is_monotone_Ioo_H1 {a b : α} {f : α → δ}
    (p1 : ContinuousOn f (Ioo a b)) (p2 : InjOn f (Ioo a b))
    {c d : α} (h2 : c ∈ Ioo a b) (h3 : d ∈ Ioo a b) (h4 : c ≤ d) :
    StrictMonoOn f (Icc c d) ∨ StrictAntiOn f (Icc c d) := by
  have h5 : Icc c d ⊆ Ioo a b := Set.Icc_subset (Ioo a b) h2 h3
  have h6 : ContinuousOn f (Icc c d) := ContinuousOn.mono p1 h5
  have h7 : InjOn f (Icc c d) := InjOn.mono h5 p2
  exact ContinuousOn.StrictMonoOn_of_InjOn'' h4 h6 h7

/-- This lemma is only used to
prove and will be immediately superseded by
`continuous_inj_is_monotone_Ioo` below.
It asserts that a continuous injective `f: (a, b) → δ` is strictly increasing if
it is strictly increasing on some closed interval `[c, d] ⊆ (a, b)`.-/
private lemma continuous_inj_is_monotone_Ioo_H2 {a b : α} {f : α → δ}
    {c d : α} (hc : c ∈ Ioo a b) (hd : d ∈ Ioo a b) (hcd : c < d)
    (p1 : ContinuousOn f (Ioo a b)) (p2 : InjOn f (Ioo a b))
    (h5 : StrictMonoOn f (Icc c d)) : StrictMonoOn f (Ioo a b) := by
  unfold StrictMonoOn
  intro x hx y hy hxy
  let s := min c x
  let t := max d y
  have hsc : s ≤ c := min_le_left c x
  have hdt : d ≤ t := le_max_left d y
  have hs : s ∈ Ioo a b := ⟨lt_min hc.1 hx.1, min_lt_of_left_lt hc.2⟩
  have ht : t ∈ Ioo a b := ⟨lt_max_of_lt_right hy.1, max_lt hd.2 hy.2⟩
  have hst : s < t := lt_of_le_of_lt hsc (lt_of_lt_of_le hcd hdt)
  have h6 : Icc c d ⊆ Icc s t := Icc_subset_Icc hsc hdt
  have h7 : StrictMonoOn f (Icc s t) ∨ StrictAntiOn f (Icc s t) :=
    continuous_inj_is_monotone_Ioo_H1 p1 p2 hs ht hst.le
  have h8' : StrictAntiOn f (Icc s t) → False := by
    intro h7
    have h9 : StrictAntiOn f (Icc c d) := StrictAntiOn.mono h7 h6
    have hcd' : c ≤ d := LT.lt.le hcd
    have hc : c ∈ Icc c d := Iff.mpr left_mem_Icc hcd'
    have hd : d ∈ Icc c d := Iff.mpr right_mem_Icc hcd'
    have hcd_1 : f c < f d := h5 hc hd hcd
    have hcd_2 : f d < f c := h9 hc hd hcd
    have : f c < f c := lt_trans hcd_1 hcd_2
    exact this.false
  have h8 : StrictMonoOn f (Icc s t) := by tauto
  have hsx : s ≤ x := min_le_right c x
  have hyt : y ≤ t := le_max_right d y
  have h9 : Icc x y ⊆ Icc s t := Icc_subset_Icc hsx hyt
  have h10 : StrictMonoOn f (Icc x y) := StrictMonoOn.mono h8 h9
  exact h10 (left_mem_Icc.mpr (le_of_lt hxy)) (right_mem_Icc.mpr (le_of_lt hxy)) hxy

/-- Every continuous injective `f : (a, b) → δ` is strictly monotone.-/
theorem continuous_inj_is_monotone_Ioo {a b : α} {f : α → δ} (h1 : a < b)
    (p1 : ContinuousOn f (Ioo a b)) (p2 : InjOn f (Ioo a b)) :
    StrictMonoOn f (Ioo a b) ∨ StrictAntiOn f (Ioo a b) := by
  have h3 : (Ioo a b).Nonempty := Iff.mpr nonempty_Ioo h1
  choose c hc using h3
  have h4 : (Ioo c b).Nonempty := Iff.mpr nonempty_Ioo hc.2
  choose d hd' using h4
  have hcd : c < d := hd'.1
  have hd : d ∈ Ioo a b := ⟨lt_trans hc.1 hcd, hd'.2⟩
  have h5 : StrictMonoOn f (Icc c d) ∨ StrictAntiOn f (Icc c d) :=
    continuous_inj_is_monotone_Ioo_H1 p1 p2 hc hd hcd.le
  have h5_1 := continuous_inj_is_monotone_Ioo_H2 hc hd hcd p1 p2
  have h5_2 (h5 : StrictAntiOn f (Icc c d)) : StrictAntiOn f (Ioo a b) := by
    let g (x : α) : OrderDual δ := f x
    have hg' : StrictMonoOn g (Ioo a b) := continuous_inj_is_monotone_Ioo_H2 hc hd hcd p1 p2 h5
    exact hg'
  tauto
```

File: Mathlib/Topology/Algebra/Order/IntermediateValue.lean (hunk 1 above)
Comment (shuxuezhuyi) [reply Mathlib/Topology/Algebra/Order/IntermediateValue.lean]: Done. Although the code is still long.

File: Mathlib/Topology/Algebra/Order/IntermediateValue.lean [hunk 2]
```diff
@@ -644,3 +648,122 @@ theorem ContinuousOn.surjOn_of_tendsto' {f : α → δ} {s : Set α} [OrdConnect
     (htop : Tendsto (fun x : s => f x) atTop atBot) : SurjOn f s univ :=
   @ContinuousOn.surjOn_of_tendsto α _ _ _ _ δᵒᵈ _ _ _ _ _ _ hs hf hbot htop
 #align continuous_on.surj_on_of_tendsto' ContinuousOn.surjOn_of_tendsto'
+
+/-- Suppose `f : [a, b] → δ` is
+continuous and injective. Then `f` is strictly monotone (increasing)
+if `f(a) < f(b)`.-/
```
Comment (jcommelin) [inline Mathlib/Topology/Algebra/Order/IntermediateValue.lean:654]:
```suggestion
if `f(a) ≤ f(b)`.-/
```

File: Mathlib/Topology/Algebra/Order/IntermediateValue.lean [hunk 3]
```diff
@@ -644,3 +648,122 @@ theorem ContinuousOn.surjOn_of_tendsto' {f : α → δ} {s : Set α} [OrdConnect
...
+      replace hat : f t < f a := this.symm.lt_of_le hat
+      have hstab : Icc s t ⊆ Icc a b := Icc_subset_Icc hs.1 ht.2
+      obtain ⟨u, hu⟩ : f a ∈ f '' Ioo s t :=
+        intermediate_value_Ioo' hst.le (hf_c.mono hstab) ⟨hat, hsa⟩
+      apply (lt_of_lt_of_le' hu.1.1 hs.1).ne'
+      exact hf_i (hstab (Ioo_subset_Icc_self hu.1)) (left_mem_Icc.mpr hab) hu.2
+
+/-- Suppose `f : [a, b] → δ` is
+continuous and injective. Then `f` is strictly antitone (decreasing)
+if `f(a) > f(b)`.-/
```
Comment (jcommelin) [inline Mathlib/Topology/Algebra/Order/IntermediateValue.lean:689]:
```suggestion
if `f(b) ≤ f(a)`.-/
```

File: Mathlib/Topology/Algebra/Order/IntermediateValue.lean (hunk 2 above)
Comment (shuxuezhuyi) [reply Mathlib/Topology/Algebra/Order/IntermediateValue.lean]: Done.

File: Mathlib/Topology/Algebra/Order/IntermediateValue.lean (hunk 3 above)
Comment (shuxuezhuyi) [reply Mathlib/Topology/Algebra/Order/IntermediateValue.lean]: Done.

File: Mathlib/Topology/Algebra/Order/IntermediateValue.lean [hunk 4]
```diff
@@ -644,3 +648,122 @@ theorem ContinuousOn.surjOn_of_tendsto' {f : α → δ} {s : Set α} [OrdConnect
...
+/-- Suppose `f : [a, b] → δ` is
+continuous and injective. Then `f` is strictly antitone (decreasing)
+if `f(b) ≤ f(a)`.-/
+theorem ContinuousOn.StrictAntiOn_of_InjOn_Icc {a b : α} {f : α → δ}
+    (hab : a ≤ b) (hfab : f b ≤ f a)
+    (hf_c : ContinuousOn f (Icc a b)) (hf_i : InjOn f (Icc a b)) :
+    StrictAntiOn f (Icc a b) := by
+    let g (x : α) : OrderDual δ := f x
+    have hgab : g a ≤ g b := hfab
+    exact ContinuousOn.StrictMonoOn_of_InjOn_Icc hab hgab hf_c hf_i
```
Comment (j-loreaux) [inline Mathlib/Topology/Algebra/Order/IntermediateValue.lean:696]: Probably this works. This is how we do it throughout the rest of the order library:
```suggestion
ContinuousOn.StrictMonoOn_of_InjOn_Icc (δ := δᵒᵈ) hab hfab hf_c hf_i
```

File: Mathlib/Topology/Algebra/Order/IntermediateValue.lean [hunk 5]
```diff
@@ -644,3 +648,122 @@ theorem ContinuousOn.surjOn_of_tendsto' {f : α → δ} {s : Set α} [OrdConnect
...
   @ContinuousOn.surjOn_of_tendsto α _ _ _ _ δᵒᵈ _ _ _ _ _ _ hs hf hbot htop
 #align continuous_on.surj_on_of_tendsto' ContinuousOn.surjOn_of_tendsto'
+
+/-- Suppose `f : [a, b] → δ` is
+continuous and injective. Then `f` is strictly monotone (increasing)
+if `f(a) ≤ f(b)`.-/
+theorem ContinuousOn.StrictMonoOn_of_InjOn_Icc {a b : α} {f : α → δ}
+    (hab : a ≤ b) (hfab : f a ≤ f b)
+    (hf_c : ContinuousOn f (Icc a b)) (hf_i : InjOn f (Icc a b)) :
+    StrictMonoOn f (Icc a b) := by
```
Comment (j-loreaux) [inline Mathlib/Topology/Algebra/Order/IntermediateValue.lean:658]: I think this theorem would be a lot less headache to prove if you did it for types first. Then you can prove the version for sets using the one for types. Note: this is how the intermediate value theorem is proved as well.

The right generality might be something like this.
```
theorem Continuous.strictMono_of_injective {α : Type*} [ConditionallyCompleteLinearOrder α] [TopologicalSpace α] [OrderTopology α] [DenselyOrdered α] {δ : Type*} [LinearOrder δ] [TopologicalSpace δ] [OrderClosedTopology δ] {f : C(α, δ)} (hf₁ : f ⊥ ≤ f ⊤) (hf₂ : Function.Injective f) : StrictMono f := sorry
```
If you do it this way, then you don't have to continue worry about whether or not your elements are in the interval or not. Then to prove the version for intervals, you just use `ContinuousOn.restrict`.

**Review by j-loreaux — COMMENTED** [review]
Primary takeaway: do things for types first then sets. Thanks for this contribution!

File: Mathlib/Topology/Algebra/Order/IntermediateValue.lean (hunk 5 above)
Comment (shuxuezhuyi) [reply Mathlib/Topology/Algebra/Order/IntermediateValue.lean:680]: Thank you very much! I add `Continuous.strictMono_of_injective` now. However, I find that the order on a closed interval is not necessarily a conditionally complete linear order. So I cannot deduce `ContinuousOn.StrictMonoOn_of_InjOn_Icc` from `Continuous.strictMono_of_injective` yet.

Comment (shuxuezhuyi) [reply Mathlib/Topology/Algebra/Order/IntermediateValue.lean:680]: I cannot deduce `ContinuousOn.StrictMonoOn_of_InjOn_Icc` from `Continuous.strictMono_of_injective` in Lean. It is very strange, and I need help: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/How.20to.20tell.20Lean.20preorders.20from.20different.20sources.20are.20same.3F

Comment (shuxuezhuyi) [reply Mathlib/Topology/Algebra/Order/IntermediateValue.lean:680]: Many thanks to @alreadydone ! I can deduce `ContinuousOn.StrictMonoOn_of_InjOn_Icc` from `Continuous.strictMono_of_injective` now.

**Comment by shuxuezhuyi** [comment]
> Primary takeaway: do things for types first then sets.

@j-loreaux Thank you very much for your advice! Now the code has been refactored.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Comment by jcommelin** [comment]
bors r-

bors d+

**Comment by jcommelin** [comment]
Let's first wait for CI to finish.

bors r-

**Pull request closed without merging**